File: naseco_fieldopsbackend/naseco_fieldopsbackend/page/season_command_centre/season_command_centre.py
import frappe
from frappe import _
from frappe.utils import cint, flt

from naseco_fieldopsbackend.fieldops_finance import get_default_company


@frappe.whitelist()
def get_command_centre(plan=None, season=None):
	plan_name = plan or _find_plan(season)
	if not plan_name:
		return {"plan": None, "message": _("Create a Season Production Plan to begin.")}

	doc = frappe.get_doc("Season Production Plan", plan_name)
	doc.check_permission("read")
	doc.refresh_actuals()
	return {
		"plan": _plan_summary(doc),
		"targets": [_target_row(row) for row in doc.production_targets],
		"stages": _stage_progress(doc.season, doc.company),
		"quality": _quality_summary(doc.season, doc.company),
		"exposure": _exposure_by_region(doc.season, doc.company),
		"resources": [row.as_dict() for row in doc.resource_allocations if row.active],
		"milestones": [row.as_dict() for row in doc.milestones],
		"readiness": [row.as_dict() for row in doc.readiness_items],
		"reports": [
			"Season Production Performance",
			"Season Grower and Acreage Progress",
			"Season Agronomy Progress",
			"Season QA Coverage",
			"Season Inspector Performance",
			"Season Input and Exposure",
			"Season Harvest and Settlement",
		],
	}


def _find_plan(season=None):
	company = get_default_company()
	if season:
		return frappe.db.get_value(
			"Season Production Plan",
			{"season": season, "company": company, "docstatus": ["<", 2]},
			"name",
			order_by="modified desc",
		)
	return frappe.db.get_value(
		"Season Production Plan",
		{"company": company, "status": ["in", ["Active", "Approved", "Under Review", "Draft"]]},
		"name",
		order_by="modified desc",
	)


def _plan_summary(doc):
	fields = (
		"name",
		"plan_title",
		"season",
		"company",
		"status",
		"calendar_status",
		"readiness_score",
		"target_outgrowers",
		"contracted_outgrowers",
		"target_plots",
		"contracted_plots",
		"target_hectares",
		"contracted_hectares",
		"planted_hectares",
		"area_achievement_percent",
		"planned_production_qty",
		"forecast_production_qty",
		"delivered_net_dry_qty",
		"production_achievement_percent",
		"qa_coverage_percent",
		"current_exposure_value",
		"assessed_harvest_value",
		"settled_net_payable",
		"deferred_bonus_liability",
		"stock_coverage_percent",
		"planned_input_cost",
	)
	return {field: doc.get(field) for field in fields}


def _target_row(row):
	return {
		"region": row.region or _("All Regions"),
		"crop": row.crop,
		"variety": row.variety,
		"production_category": row.production_category,
		"target_outgrowers": cint(row.target_outgrowers),
		"target_hectares": flt(row.target_hectares),
		"planned_production_qty": flt(row.planned_production_qty),
	}


def _stage_progress(season, company):
	return frappe.db.sql(
		"""
		select
			coalesce(stage.stage_name, 'Not Started') stage,
			count(*) crop_cycles,
			coalesce(sum(contract.contracted_area_hectares), 0) hectares
		from `tabCrop Cycle` cycle
		left join `tabCrop Cycle Stage` stage on stage.name = cycle.current_stage
		left join `tabOutgrower Production Contract` contract
			on contract.name = cycle.production_contract
		where cycle.season = %(season)s and cycle.company = %(company)s
		group by coalesce(stage.stage_name, 'Not Started'), coalesce(stage.order_index, 0)
		order by coalesce(stage.order_index, 0)
		""",
		{"season": season, "company": company},
		as_dict=True,
	)


def _quality_summary(season, company):
	return frappe.db.sql(
		"""
		select
			status,
			count(*) inspection_count,
			round(avg(farmer_compliance_percent), 1) farmer_compliance,
			round(avg(supervisor_compliance_percent), 1) supervisor_compliance,
			round(avg(spacing_compliance_percent), 1) spacing_compliance,
			round(avg(average_gps_accuracy_m), 1) gps_accuracy
		from `tabInspection` inspection
		where inspection.season = %(season)s
			and exists (
				select 1 from `tabCrop Cycle` cycle
				where cycle.name = inspection.crop_cycle and cycle.company = %(company)s
			)
		group by status
		order by field(status, 'Awaiting QA Review', 'Reinspection Required',
			'In Progress', 'Scheduled', 'Verified', 'Cancelled')
		""",
		{"season": season, "company": company},
		as_dict=True,
	)


def _exposure_by_region(season, company):
	return frappe.db.sql(
		"""
		select
			coalesce(outgrower.region, 'Unassigned') region,
			count(*) crop_cycles,
			coalesce(sum(cycle.recoverable_stock_value), 0) stock_inputs,
			coalesce(sum(cycle.cash_advanced), 0) cash_advances,
			coalesce(sum(cycle.total_exposure), 0) total_exposure,
			coalesce(sum(cycle.forecast_net_payable), 0) forecast_net_payable
		from `tabCrop Cycle` cycle
		left join `tabFarm Plot` plot on plot.name = cycle.plot
		left join `tabOutgrower` outgrower on outgrower.name = plot.outgrower
		where cycle.season = %(season)s and cycle.company = %(company)s
		group by coalesce(outgrower.region, 'Unassigned')
		order by total_exposure desc
		""",
		{"season": season, "company": company},
		as_dict=True,
	)

File: naseco_fieldopsbackend/naseco_fieldopsbackend/page/season_command_centre/season_command_centre.js
frappe.pages['season-command-centre'].on_page_load = function(wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Season Command Centre'),
		single_column: true
	});
	const state = { page, wrapper, plan: frappe.route_options?.plan || null };
	frappe.route_options = null;
	page.add_field({
		fieldname: 'season',
		label: __('Season'),
		fieldtype: 'Link',
		options: 'Season',
		change() {
			load_command_centre(state, this.get_value());
		}
	});
	page.set_primary_action(__('Refresh'), () => load_command_centre(state), 'refresh');
	$(wrapper).find('.layout-main-section').html('<div class="season-command-centre"></div>');
	load_command_centre(state);
};

function load_command_centre(state, season) {
	frappe.call({
		method: 'naseco_fieldopsbackend.naseco_fieldopsbackend.page.season_command_centre.season_command_centre.get_command_centre',
		args: { plan: state.plan, season },
		freeze: true,
		callback(r) {
			const data = r.message || {};
			state.plan = data.plan && data.plan.name;
			render_command_centre(state, data);
		}
	});
}

function render_command_centre(state, data) {
	const root = $(state.wrapper).find('.season-command-centre');
	if (!data.plan) {
		root.html(`<div class="cc-empty">${escape_html(data.message || __('No production plan found.'))}</div>`);
		return;
	}
	const p = data.plan;
	state.page.set_title_sub(`${p.season} · ${p.company}`);
	state.page.clear_inner_toolbar();
	state.page.add_inner_button(__('Open Production Plan'), () => {
		frappe.set_route('Form', 'Season Production Plan', p.name);
	});

	root.html(`
		<div class="cc-heading">
			<div>
				<h2>${escape_html(p.plan_title || p.name)}</h2>
				<div class="text-muted">${escape_html(p.season)} · ${escape_html(p.calendar_status || '')}</div>
			</div>
			<span class="indicator-pill ${status_color(p.status)}">${escape_html(p.status)}</span>
		</div>
		<div class="cc-metrics">
			${metric(__('Readiness'), p.readiness_score, '%')}
			${metric(__('Contracted Hectares'), p.contracted_hectares, '', p.target_hectares)}
			${metric(__('Planted Hectares'), p.planted_hectares, '', p.target_hectares)}
			${metric(__('Production Forecast'), p.forecast_production_qty, ' kg', p.planned_production_qty)}
			${metric(__('QA Coverage'), p.qa_coverage_percent, '%')}
			${currency_metric(__('Current Exposure'), p.current_exposure_value)}
			${currency_metric(__('Assessed Harvest'), p.assessed_harvest_value)}
			${currency_metric(__('Settled Payable'), p.settled_net_payable)}
		</div>
		<div class="cc-grid">
			${panel(__('Crop Stage Progress'), table(
				[__('Stage'), __('Crop Cycles'), __('Hectares')],
				(data.stages || []).map(row => [
					row.stage, format_number(row.crop_cycles), format_number(row.hectares, null, 2)
				])
			))}
			${panel(__('Quality Assurance'), table(
				[__('Status'), __('Inspections'), __('Farmer %'), __('Supervisor %'), __('Spacing %'), __('GPS m')],
				(data.quality || []).map(row => [
					row.status, format_number(row.inspection_count),
					format_number(row.farmer_compliance, null, 1),
					format_number(row.supervisor_compliance, null, 1),
					format_number(row.spacing_compliance, null, 1),
					format_number(row.gps_accuracy, null, 1)
				])
			))}
			${panel(__('Exposure by Region'), table(
				[__('Region'), __('Cycles'), __('Stock Inputs'), __('Cash Advances'), __('Exposure')],
				(data.exposure || []).map(row => [
					row.region, format_number(row.crop_cycles),
					format_currency(row.stock_inputs), format_currency(row.cash_advances),
					format_currency(row.total_exposure)
				])
			))}
			${panel(__('Readiness Controls'), table(
				[__('Area'), __('Owner'), __('Status')],
				(data.readiness || []).map(row => [
					row.readiness_area, row.responsible_user || '-', row.status
				])
			))}
			${panel(__('Season Milestones'), table(
				[__('Milestone'), __('Planned End'), __('Status')],
				(data.milestones || []).map(row => [
					row.milestone, frappe.datetime.str_to_user(row.planned_end_date), row.status
				])
			))}
			${panel(__('Management Reports'), `
				<div class="cc-report-list">
					${(data.reports || []).map(report => `
						<button class="btn btn-default btn-sm cc-report" data-report="${escape_html(report)}">
							${escape_html(report)}
						</button>
					`).join('')}
				</div>
			`)}
		</div>
	`);
	root.find('.cc-report').on('click', function() {
		frappe.set_route('query-report', $(this).data('report'), { season: p.season, company: p.company });
	});
}

function metric(label, value, suffix = '', target = null) {
	const target_text = target === null || target === undefined
		? ''
		: `<div class="cc-target">${__('Target')} ${format_number(target, null, 1)}</div>`;
	return `<div class="cc-metric">
		<div class="cc-label">${escape_html(label)}</div>
		<div class="cc-value">${format_number(value || 0, null, 1)}${suffix}</div>
		${target_text}
	</div>`;
}

function currency_metric(label, value) {
	return `<div class="cc-metric">
		<div class="cc-label">${escape_html(label)}</div>
		<div class="cc-value cc-currency">${format_currency(value || 0)}</div>
	</div>`;
}

function panel(title, content) {
	return `<section class="cc-panel"><h3>${escape_html(title)}</h3>${content}</section>`;
}

function table(headers, rows) {
	return `<div class="table-responsive"><table class="table table-hover">
		<thead><tr>${headers.map(h => `<th>${escape_html(h)}</th>`).join('')}</tr></thead>
		<tbody>${rows.length ? rows.map(row => `<tr>${row.map(value => `<td>${escape_html(value ?? '-')}</td>`).join('')}</tr>`).join('') : `<tr><td colspan="${headers.length}" class="text-muted">${__('No records')}</td></tr>`}</tbody>
	</table></div>`;
}

function status_color(status) {
	if (['Active', 'Approved', 'Closed'].includes(status)) return 'green';
	if (['Awaiting Approval', 'Under Review', 'Closing'].includes(status)) return 'orange';
	if (status === 'Cancelled') return 'red';
	return 'gray';
}

function escape_html(value) {
	return frappe.utils.escape_html(String(value ?? ''));
}
